File: Data Structures & Algorithms/delete-node-in-a-bst/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} key
     * @return {TreeNode}
     */
    findMinNode(root) {
        let curr = root;
        while (curr && curr.left) {
            curr = curr.left;
        }

        return curr;
    }
    deleteNode(root, key) {
        if (!root) return null;

        if (key > root.val) {
            root.right = this.deleteNode(root.right, key);
        } else if (key < root.val) {
            root.left = this.deleteNode(root.left, key);
        } else {
            if (!root.left) return root.right;
            if (!root.right) return root.left;

            const minNode = this.findMinNode(root.right);
            root.val = minNode.val;
            root.right = this.deleteNode(root.right, minNode.val);
        }

        return root;
    }
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        queue = []
        res = []

        if root:
            queue.append(root)

        while len(queue) > 0:
            level_length = len(queue)

            for i in range(level_length):
                curr = queue.pop(0)

                if i == level_length - 1:
                    res.append(curr.val)

                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
        return res
